### PR TITLE
[PIN-2387] Added route needed to update agreement in draft

### PR DESCRIPTION
**Pull request by bloodsky**

### CODE REVIEW

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala [hunk 1]
```diff
@@ -196,6 +196,40 @@ final case class AgreementApiServiceImpl(
...
+    * Code: 200, Message: Agreement updated., DataType: Agreement
+    * Code: 404, Message: Agreement not found, DataType: Problem
+    * Code: 400, Message: Invalid ID supplied, DataType: Problem
+    */
+  override def updateAgreementById(agreementId: String, agreementUpdatePayload: AgreementUpdatePayload)(implicit
+    contexts: Seq[(String, String)],
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem],
+    toEntityMarshallerAgreement: ToEntityMarshaller[Agreement]
+  ): Route = {
+    val result = for {
```
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala]: Explicit the type please

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala [hunk 2]
```diff
@@ -196,6 +196,40 @@ final case class AgreementApiServiceImpl(
...
+      agreementUUID <- agreementId.toFutureUUID
+      agreement     <- agreementManagementService.getAgreementById(agreementUUID)
+      seed = AgreementManagement.UpdateAgreementSeed(
```
Comment (beetlecrunch) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala:216]: This update must be granted only to draft agreements, so I suggest you add a control here.

File: src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        )
+      )
+
+      /* | Maybe a trivial test ? |
+      val payload = AgreementUpdatePayload("consumer-notes")
+      Get() ~> service.updateAgreementById(agreement.id.toString, payload) ~> check {
+              status shouldEqual StatusCodes.InternalServerError
+      }*/
+
+      Get() ~> service.upgradeAgreementById(agreement.id.toString) ~> check {
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala]:
```suggestion
      Get() ~> service.updateAgreementById(agreement.id.toString, <payloadHere>) ~> check {
```

File: src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+          eserviceId = eService.id,
+          descriptorId = newerDescriptor.id,
+          consumerId = consumer.id
+        )
+
+      val seed = UpgradeAgreementSeed(descriptorId = newerDescriptor.id, SpecData.defaultStamp.get)
+
+      mockAgreementRetrieve(agreement)
+      mockEServiceRetrieve(eService.id, eService)
+      mockAgreementUpgrade(agreement.id, seed, newAgreement)
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala]:
```suggestion
      mockAgreementUpdate(agreement.id, seed, newAgreement)
```
and this should not return a new agreement, but the same agreement with the updated fields

File: src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        )
+      val newAgreement      =
+        agreement.copy(
+          id = UUID.randomUUID(),
+          eserviceId = eService.id,
+          descriptorId = newerDescriptor.id,
+          consumerId = consumer.id
+        )
+
+      val seed = UpgradeAgreementSeed(descriptorId = newerDescriptor.id, SpecData.defaultStamp.get)
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala]:
```suggestion
      val seed = AgreementUpdatePayload(...)
```

File: src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+import java.util.UUID
+
+class AgreementUpdateSpec extends AnyWordSpecLike with SpecHelper with ScalatestRouteTest {
+
+  import agreementApiMarshaller._
+
+  "Agreement update" should {
+    "succedd" in {
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala]:
```suggestion
    "succeded" in {
```

File: src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  ClientComponentState
+}
+
+import java.util.UUID
+
+class AgreementUpdateSpec extends AnyWordSpecLike with SpecHelper with ScalatestRouteTest {
+
+  import agreementApiMarshaller._
+
+  "Agreement update" should {
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala]: let's add also a failure test where the given agreement is not in draft

**Review by galales — CHANGES_REQUESTED** [review]

File: src/main/scala/it/pagopa/interop/agreementprocess/common/Adapters.scala [hunk 8]
```diff
@@ -16,6 +16,8 @@ object Adapters {
   val ARCHIVABLE_STATES: Set[AgreementManagement.AgreementState]  =
     Set(AgreementManagement.AgreementState.ACTIVE, AgreementManagement.AgreementState.SUSPENDED)
   val SUBMITTABLE_STATES: Set[AgreementManagement.AgreementState] = Set(AgreementManagement.AgreementState.DRAFT)
+  val UPDATEABLE_STATES: Set[AgreementManagement.AgreementState]  =
+    Set(AgreementManagement.AgreementState.DRAFT)
```
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/agreementprocess/common/Adapters.scala:20]: 🔪

File: src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+import org.scalatest.matchers.should.Matchers._
+import org.scalatest.wordspec.AnyWordSpecLike
+
+import java.util.UUID
+
+class AgreementUpdateSpec extends AnyWordSpecLike with SpecHelper with ScalatestRouteTest {
+
+  import agreementApiMarshaller._
+
+  "Agreement Update" should {
```
Comment (galales) [inline src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala:16]: to keep the same approach and improve tests results readability I suggest to move tests under the same parent, when it is the same.
In this case 
```
"Agreement Update" should {
  "succeed if agreement is in draft" in {...}
  "fail if agreement is not in draft" in {...}
  "fail if requester is not the producerId" in {...}
}
```

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala [hunk 10]
```diff
@@ -196,6 +196,44 @@ final case class AgreementApiServiceImpl(
     }
   }
 
+  /**
+    * Code: 200, Message: Agreement updated., DataType: Agreement
+    * Code: 404, Message: Agreement not found, DataType: Problem
+    * Code: 400, Message: Invalid ID supplied, DataType: Problem
+    */
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala:203]: I suggest to remove these auto generated comments, they are rarely maintained and then become misleading

**Review by galales — APPROVED** [review]

File: src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+          consumerId = consumerId,
+          producerId = requesterOrgId,
+          consumerNotes = Some("old-consumer-notes")
+        )
+
+      val updatedAgreement =
+        agreement.copy(consumerNotes = Some(payload.consumerNotes))
+
+      val seed = UpdateAgreementSeed(
+        state = agreement.state,
```
Comment (TonioGela) [inline src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala:38]: is the default state DRAFT? it's not visible here, may you force it using `state = DRAFT`?

File: src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+      mockAgreementUpdate(agreement.id, seed, updatedAgreement)
+
+      Get() ~> service.updateAgreementById(agreement.id.toString, payload) ~> check {
+        status shouldEqual StatusCodes.OK
+      }
+    }
+  }
+
+  "Agreement Update" should {
+    "fail if agreement is not in draft" in {
```
Comment (TonioGela) [inline src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala:59]: Same here, how can I tell the agreement is not in draft state?

Comment (bloodsky) [reply src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala:59]: You can tell that this agreement is not in draft state from this line (using SpecData)
`val agreement = SpecData.activeAgreement.copy(...)`

File: src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala (hunk 11 above)
Comment (bloodsky) [reply src/test/scala/it/pagopa/interop/agreementprocess/AgreementUpdateSpec.scala:38]: I can force it but the agreement in this case is specified to be in in draft state (using SpecData)
`val agreement = SpecData.draftAgreement.copy(...)`

**Review by beetlecrunch — APPROVED** [review]

**Pull request merged**